java-checks/src/main/java/org/sonar/java/checks/SpecializedFunctionalInterfacesCheck.java

```diff
@@ -0,0 +1,239 @@
+        case "java.util.function.Function":
+          reportString = handleFunctionInterface(arguments);
+          break;
+        case "java.util.function.BiFunction":
+          reportString = handleBiFunctionInterface(arguments);
+          break;
+        case "java.util.function.BiConsumer":
+          reportString = handleBiConsumerInterface(arguments);
+          break;
+        case "java.util.function.Supplier":
+        case "java.util.function.Consumer":
+        case "java.util.function.Predicate":
+        case "java.util.function.UnaryOperator":
+        case "java.util.function.BinaryOperator":
+          reportString = handleOneParameterInterface(arguments).map(s -> s + interfaceType.symbolType().name());
+          break;
+        default:
+          return Optional.empty();
+      }
+      if (reportString.isPresent()) {
+        return Optional.of(new InterfaceTreeAndStringPairReport(reportString.get(), interfaceType));
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<String> handleOneParameterInterface(TypeArguments args) {
+    String reportString = new ParameterTypeNameAndTreeType(args.get(0)).paramTypeName;
+    if (reportString != null) {
+      return Optional.of(reportString);
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<String> handleFunctionInterface(TypeArguments args) {
+    ParameterTypeNameAndTreeType firstArgument = new ParameterTypeNameAndTreeType(args.get(0));
+    ParameterTypeNameAndTreeType secondArgument = new ParameterTypeNameAndTreeType(args.get(1));
+    String reportString = null;
+    if (firstArgument.paramType.equals(secondArgument.paramType)) {
+      reportString = "UnaryOperator<" + firstArgument.paramType.name() + ">";
+    } else if (firstArgument.paramTypeName != null && secondArgument.paramTypeName != null) {
+      reportString = firstArgument.paramTypeName + "To" + secondArgument.paramTypeName + "Function";
+    } else if (firstArgument.paramTypeName == null && secondArgument.paramTypeName != null) {
+      reportString = "To" + secondArgument.paramTypeName + "Function<" + firstArgument.paramType.name() + ">";
+    } else if (firstArgument.paramTypeName != null) {
+      reportString = firstArgument.paramTypeName + "Function<" + secondArgument.paramType.name() + ">";
+    }
+    if (reportString != null) {
+      return Optional.of(reportString);
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<String> handleBiFunctionInterface(TypeArguments args) {
+    ParameterTypeNameAndTreeType firstArgument = new ParameterTypeNameAndTreeType(args.get(0));
+    ParameterTypeNameAndTreeType secondArgument = new ParameterTypeNameAndTreeType(args.get(1));
+    ParameterTypeNameAndTreeType thirdArgument = new ParameterTypeNameAndTreeType(args.get(2));
+    if (firstArgument.paramType.equals(secondArgument.paramType) && firstArgument.paramType.equals(thirdArgument.paramType)) {
```

Comment:
How do you know that `firstArgument.paramType` is not going to be null ?! I see a NPE hidden here.

Reply:
I think that this won't happen, so no reason to be nullable.

Reply:
```
BiFunction<Integer[][][], Integer[][][], Integer[][][]> myBiFunc = new BiFunction<Integer[][][], Integer[][][], Integer[][][]>() {
      @Override
      public Integer[][][] apply(Integer[][][] t, Integer[][][] u) {
        return null;
      }
    };
```